### ISSUE TITLE
JsonCreator not working on Classes with JsonTypeInfo when input string is missing the type field

### ISSUE THREAD

**Issue by Ragsboss** [FasterXML/jackson-databind#747]
``` java
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type", include = JsonTypeInfo.As.PROPERTY)
@JsonSubTypes({
      @Type(value = IntegerLiteral.class, name = "integer"),
      @Type(value = StringLiteral.class, name = "string"),
})
public abstract class Literal {
  public abstract Object getValue();

  @JsonCreator
  public static IntegerLiteral createInt(@JsonProperty("value") long lVal) {
      return new IntegerLiteral(lVal);
  }
}

public class IntegerLiteral extends Literal {
  protected long value;

  @Override
  public Long getValue() {
     return value;
  }

  public IntegerLiteral(long value) {
      this.value = value;
  }
}

public class Field {
  private Literal value;

  public Literal getValue { return value; }
  public void setValue(Literal val) { value = val; }
}
```

My goal is to support both of the following json to represent an instance of Field class

{ "value": 4 } and

{ "value": { "type": "integer", "value": 4 } }

The second one works, but the first one fails with below exception.

I CANNOT remove JsonSubTypes or JsonTypeInfo annotations on Literal class - that's my constraint. I googled a lot to see how I can make jackson give my JsonCreator precedence over JsonTypeInfo - but to no avail. I tried with and without JsonProperty annotation on the JsonCreator method.

``` text
com.fasterxml.jackson.databind.JsonMappingException: Unexpected token (VALUE_NUMBER_INT), expected FIELD_NAME: missing property 'type' that is to contain type id  (for class Literal)
 at [Source: java.io.StringReader@45002203; line: 1, column: 2] (through reference chain: Field["value"])
    at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:164) ~[jackson-databind-2.1.2.jar:2.1.2]
    at com.fasterxml.jackson.databind.DeserializationContext.wrongTokenException(DeserializationContext.java:692) ~[jackson-databind-2.1.2.jar:2.1.2]
    at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedUsingDefaultImpl(AsPropertyTypeDeserializer.java:140) ~[jackson-databind-2.1.2.jar:2.1.2]
    at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromObject(AsPropertyTypeDeserializer.java:73) ~[jackson-databind-2.1.2.jar:2.1.2]
    at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeWithType(BeanDeserializerBase.java:850) ~[jackson-databind-2.1.2.jar:2.1.2]
    at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:373) ~[jackson-databind-2.1.2.jar:2.1.2]
    at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:98) ~[jackson-databind-2.1.2.jar:2.1.2]
    at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:308) ~[jackson-databind-2.1.2.jar:2.1.2]
    at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:121) ~[jackson-databind-2.1.2.jar:2.1.2]
    at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:2797) ~[jackson-databind-2.1.2.jar:2.1.2]
    at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:1943) ~[jackson-databind-2.1.2.jar:2.1.2]
```

**Comment by cowtowncoder** [comment]
Right, unfortunately there is no way to do that. Object id and Type id handling must precede processing first, and only once those are resolved would creator get involved.
This is by design. Error messaging may be improved and I am open to suggestions there, but I don't think it is possible to bend existing object, type id handling functionality to work the way you want.

Instead you will probably need to implement custom deserializers to support such structure.

**Comment by Ragsboss** [comment]
Thank you for looking into this, appreciate it. Is there any other (dynamic) means to supply type info? Short of custom deserializer is there any other extensibility point we can make use of, to achieve this?

If none, we'll have to likely consider custom deserializer - but here, do we have a way to register custom deserializers for specific java types as opposed to taking the full ownership of entire deserialization process?

**Comment by cowtowncoder** [comment]
Yes, you can register per-type (de)serializers. In fact, via `Module`, you define `Serializers` / `Deserializers`, which gets called for given type, so you can use whatever criteria you want.

Or, alternatively, what sometimes makes more sense is to register `BeanSerializerModifier` / `BeanDeserializerModifier`, which gives access to modify POJO (de)serializers while they are being constructed. So you can, for example, take default (de)serializer, create your own wrapper, and either handle process (for that instance) or delegate, on case-by-case basis.

Also, default (de)serializers use delegation so there are mechanisms for your (de)serializer to handle particular type but then delegate for contained/referenced types (properties of a POJO).

So you do not need to take ownership of the full chain. Of course registrations and handling is not trivial so ideally custom handlers should only be used if other options are not applicable.

**Comment by cowtowncoder** [comment]
As to type info. Yes, it is possible to also register Type Id resolvers (`TypeIdResolver`), builder for resolver (`TypeResolverBuilder`) and so on. One (latter) for extracting out type id to use, another for resolving from String type id into actual `Class`. I am not sure if it would be possible to combine these to allow you to support your use case, but it may be worth the try. It's not that structure must match as much as that delegation works in certain way: type handlers "wrap" regular (de)serialization.
You can see the delegation model via `JsonSerializer` and `JsonDeserializer`; it is somewhat different for different cases (for serialization it's simpler as it just writes bit more info; for deserialization needs to drive type discovery more).

**Comment by scottrw93** [comment]
I would also like to be able to do this. 

Would it not make sense to fallback to the deserilzer for the basetype if the defaultImpl is not specified in the annotation? Would seem like this could be done [here](https://github.com/FasterXML/jackson-databind/blob/fbc1fa23c92a69da474f3a7039e5494de7dc7ae2/src/main/java/com/fasterxml/jackson/databind/jsontype/impl/AsPropertyTypeDeserializer.java#L134) with this snippet

```java    
    @SuppressWarnings("resource")
      protected Object _deserializeTypedUsingDefaultImpl(JsonParser p, DeserializationContext ctxt,
            TokenBuffer tb) throws IOException
    {
        // As per [JACKSON-614], may have default implementation to use
        JsonDeserializer<Object> deser = _findDefaultImplDeserializer(ctxt);
        if(deser == null) {
            //fallback to getting deserializer for basetype
            deser =  _findDeserializer(ctxt, super.baseTypeName());
        }
        if (deser != null) {
            if (tb != null) {
                tb.writeEndObject();
                p = tb.asParser(p);
                // must move to point to the first token:
                p.nextToken();
            }
            return deser.deserialize(p, ctxt);
        }
        // or, perhaps we just bumped into a "natural" value (boolean/int/double/String)?
        Object result = TypeDeserializer.deserializeIfNatural(p, ctxt, _baseType);
        if (result != null) {
            return result;
        }
        // or, something for which "as-property" won't work, changed into "wrapper-array" type:
        if (p.getCurrentToken() == JsonToken.START_ARRAY) {
            return super.deserializeTypedFromAny(p, ctxt);
        }
        ctxt.reportWrongTokenException(p, JsonToken.FIELD_NAME,
                "missing property '"+_typePropertyName+"' that is to contain type id  (for class "+baseTypeName()+")");
        return null;
    }
```

**Comment by cowtowncoder** [comment]
@Ragsboss actually perhaps specifying `defaultImpl = IntegerLiteral.class` should work here.

@scottrw93 I don't know. I am bit hesitant to add new kinds of fallbacks for case where missing information may well signal a problem elsewhere -- in general type id is expected to be used; and if not, `defaultImpl` indicates that this is acceptable. So wouldn't it make sense to specify that `defaultImpl` explicitly?

**Comment by scottrw93** [comment]
The problem is if you have a class that extends a class that has that annotation, the default impl will not be valid for that extended class and will fail to deserialise. You could just put the annotation on to the extended class as well, but in my use case lots of classes will extend it and it is cleaner not to need a new annotation with a different default impl on all of them.

**Comment by cowtowncoder** [comment]
@scottrw93 Such usage is incorrect: all subtypes must have conceptually same shared base class. Their default implementations can not vary either. Leaving out `defaultImpl` would not necessarily help either since "default default" would then be contextual base type.

At this point I will close this issue since original usage is incorrect and assumes behavior that is not what Jackson provides. Type id must be included, except if `defaultImpl` is specified. If it is specified, deserialization should work and there is nothing specific about `@JsonCreator`: it is applied only after polymorphic type resolution has completed. It does not compete with `@JsonTypeInfo` in any way and specifically can not "override" it. Processing starts with handling of excepted Object Id (if any); followed by Type Id (if any); finally delegating to the deserializer for resolved type.

@scottrw93 If you would like to request additional defaulting above and beyond `defaultImpl`, please file a separate issue -- it is usually better to have clear lineage, and your request is somewhat different from problem as reported originally here (it is related, yes, but not same thing). I am open to changes but just want to be clear on what exactly is being suggested.
